# Question1

## 1.1 Replication the regression table

In [16]:
pwd

'/Users/anhnguyen/Documents/GitHub/2025AAE722_AnhNguyen'

In [15]:
import pandas as pd
import csv

df = pd.read_csv('Advertising.csv')  # replace with your file path
print(df.head())

# Count the number of rows and columns
num_rows, num_cols = df.shape
print(f'Number of rows: {num_rows}, Number of columns: {num_cols}')
# List all column names
print(f'Column names: {df.columns.tolist()}')
# Display data types of each column
print(f'Data types:\n{df.dtypes}')

   Unnamed: 0     TV  radio  newspaper  sales
0           1  230.1   37.8       69.2   22.1
1           2   44.5   39.3       45.1   10.4
2           3   17.2   45.9       69.3    9.3
3           4  151.5   41.3       58.5   18.5
4           5  180.8   10.8       58.4   12.9
Number of rows: 200, Number of columns: 5
Column names: ['Unnamed: 0', 'TV', 'radio', 'newspaper', 'sales']
Data types:
Unnamed: 0      int64
TV            float64
radio         float64
newspaper     float64
sales         float64
dtype: object


In [19]:
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np
import pandas as pd

# Fit a multiple linear regression of sales on TV, radio, newspaper
import statsmodels.api as sm

# use existing dataframe `df`
X = df[['TV', 'radio', 'newspaper']]
y = df['sales']

# 1) Statsmodels OLS (full summary)
X_const = sm.add_constant(X)
ols_model = sm.OLS(y, X_const).fit()
print(ols_model.summary())



                            OLS Regression Results                            
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Sat, 20 Sep 2025   Prob (F-statistic):           1.58e-96
Time:                        22:43:17   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9389      0.312      9.422      0.0

## 1.2 Hypothese 

a. 

Null Hypothesis (H0): There is no relationship between TV advertising and Sales.
   
    H0: Beta1 = 0

If Beta1=0, the estimated regression model reduces to: ' estiamted Sales' = 2.939 + 0.189 X Radio - 0.001 X Newspaper, meaning TV does not contribute to explaining Sales.
    
Alternative Hypothesis (HA): There is a relationship between TV advertising and Sales.
    
    H0: Beta1 != 0

b. 

Null Hypothesis (H0): There is no relationship between Radio advertising and Sales.
   
    H0: Beta2 = 0

If Beta2=0, the estimated regression model reduces to: ' estiamted Sales' = 2.939 + 0.046 X TV - 0.001 X Newspaper, meaning Radio does not contribute to explaining Sales.
    
Alternative Hypothesis (HA): There is a relationship between Radio advertising and Sales.
    
    H0: Beta2 != 0

c. 

Null Hypothesis (H0): There is no relationship between Newspaper advertising and Sales.
   
    H0: Beta3 = 0

If Beta3=0, the estimated regression model reduces to: ' estiamted Sales' = 2.939 + 0.046 X TV + 0.189 X Radio, meaning Newspaper does not contribute to explaining Sales.
    
Alternative Hypothesis (HA): There is a relationship between Newspaper advertising and Sales.
    
    H0: Beta3 != 0
    







## 1.3 Interpreting the results





a. Overall model is statistically significant (F-test p = 1.58e-96). At least one of the predictors in the model explains variation in sales (the dependent variable).


b. TV: coef = 0.046, p = 1.51e-81 -> statistically significant at 5%.
Interpretation: a one-unit increase in TV is associated with an average 0.046 change in sales (holding other predictors constant).


c. Radio: coef = 0.189, p = 1.51e-54 -> statistically significant at 5%. 
Interpretation: a one-unit increase in radio is associated with an average 0.189 change in sales (holding other predictors constant).


d. Newspaper: coef = -0.001, p = 0.86 -> not statistically significant at 5%. Interpretation: no strong evidence that newspaper affects sales once other predictors are controlled for.


# Question 2


## Concepts

a. K-Nearest Neighbors (KNN):

KNN is a non-parametric, instance-based learning method. Instead of assuming a functional form, KNN uses the proximity of data points in feature space to make predictions.

Some common distance metrics are Euclidean, Manhattan, Minkowski, or cosine similarity.

b. KNN Classifier:

KNN Classifier helps to predict a categorial label.

We choose a value of K (the number of neighbors to consider). After KNN Classifier calculates the distance from a new observation to all training points, it will identify the K nearest neighbors and take a majority vote to classify the new observation to the class most common among the K neighbors.

c. KNN Regression:

KNN Regression helps to predict a continuous numeric value. 

We choose a value of K. KNN Regression finds its K nearest neighbors after calculating the distance from a new observatin to all training points. The outcome is the weighted average of the target values. 



## Differences

a. Outputs for KNN Classifier is discrete labels while KNN Regression's outputs are the continuous.

b. Decision rule for KNN Classifier is the majority voting among neighbors, while KNN Regression's decision is based on average or weighted avearage of neighbors. 

c. KNN Classifier evaluates models based on accuracy, precision, recall, and F1 score metrics. KNN Regression evaluates models based on R-squared, RMSE, MAE.

d. Loss fucntions (how wrong a model's prediction is) for KNN Classifier are classification error and cross-entropy. Loss functions for KNN Regression are mean squared error (MSE), mean absolute error (MAE).


# Question 3


## (a) Training RSS - Linear true relationship

The cublic model training's residual sum of squares (RSS) will be lower than the linear model training's.

This is because the linear model is nested inside the cubic model. The cubic model is more flexible, fits more noise, thus achieves a lower training RSS. 

## (b) Test RSS - Linear true relationship

The linear model will have a lower expected test RSS when the true relationship is linear. 

This is because of bias-variance tradeoff issue. The linear model is correctly specified, achieving lower/zero bias and lower variance. The cubic model is more flexible, thus increasing variance without reducing bias and overfitting the training noise, making cubic model generalize worse than the linear model. 

## (c) Training RSS - Non-linear true relationship

The cubmic model training's RSS will be lower than the linear model training's. 

This is because the linear model is nested inside the cublic model. In OLS, adding regressors cannot increase in-sample fit. Adding regressors into an OLS function is usually stricter unless the related coefficients land exactly at 0. 

## (d) Test RSS - Non-linear true relationship

There is not enough information to tell which model has lower RSS. 

With a nonlinear true relationship, the cubic can reduce bias relative to the linear, but it also adds variance (two extra parameters) and may still be misspecified if the true function isn’t well-approximated by a cubic. Which effect dominates depends on: (1) how nonlinear the truth is, (2) noise level, (3) sample size, (4) the predictor's distribution.

In this case, it is practical to use cross-validation test.



# Question 4


In [16]:
import statsmodels.formula.api as smf
import numpy as np
import pandas as pd



# Generate reproducible data
rng = np.random.default_rng(1)
x = rng.normal(size=100)
y = 2 * x + rng.normal(size=100)


## (a) Y onto X regression (without an intercept)

In [19]:
# Fit OLS without an intercept

# reshape x to 2D (n x 1)
x = x.reshape(-1, 1)

# Fit OLS without intercept
model_no_intercept = sm.OLS(y, x).fit()

beta = model_no_intercept.params[0]
se = model_no_intercept.bse[0]
tstat = model_no_intercept.tvalues[0]
pval = model_no_intercept.pvalues[0]

print(model_no_intercept.summary())


                                 OLS Regression Results                                
Dep. Variable:                      y   R-squared (uncentered):                   0.743
Model:                            OLS   Adj. R-squared (uncentered):              0.740
Method:                 Least Squares   F-statistic:                              285.6
Date:                Sun, 21 Sep 2025   Prob (F-statistic):                    6.23e-31
Time:                        12:36:33   Log-Likelihood:                         -141.35
No. Observations:                 100   AIC:                                      284.7
Df Residuals:                      99   BIC:                                      287.3
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

### Comments

(a) The coefficient estimate is very close to the true value of 2 used in the data-generating process.

(b) The standard error is small, indicating a precise estimate.

(c) The t-statistic tests the null hypothesis that the coefficient is zero. The t-statistic is large in magnitude, suggesting we can reject the null hypothesis.

(d) The p-value indicates the significance of the coefficient. The p-value is very small, suggesting strong evidence against the null hypothesis.
    
(e) The R-squared value indicates that a large proportion of the variance in Y is explained by X.
    
Overall, the results suggest a strong positive relationship between X and Y, consistent with the data-generating process.

## (b) X onto Y regression (without an intercept)

In [29]:
x = x.reshape(-1, 1)   # predictor must be 2D
model_xy = sm.OLS(x, y).fit()   # x ~ β*y, no intercept

beta_xy = model_xy.params[0]
se_xy   = model_xy.bse[0]
tstat_xy= model_xy.tvalues[0]
pval_xy = model_xy.pvalues[0]

print(model_xy.summary())


                                 OLS Regression Results                                
Dep. Variable:                      y   R-squared (uncentered):                   0.743
Model:                            OLS   Adj. R-squared (uncentered):              0.740
Method:                 Least Squares   F-statistic:                              285.6
Date:                Sun, 21 Sep 2025   Prob (F-statistic):                    6.23e-31
Time:                        12:50:41   Log-Likelihood:                         -58.349
No. Observations:                 100   AIC:                                      118.7
Df Residuals:                      99   BIC:                                      121.3
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

### Comments

(a) The coefficient estimate is 0.4, close to the true value of 1/2 (or 0.5) used in the data-generating process.

(b) The standard error is small, indicating a precise estimate.

(c) The t-statistic tests the null hypothesis that the coefficient is zero. The t-statistic is 16.89, suggesting we can reject the null hypothesis.

(d) The p-value indicates the significance of the coefficient. The p-value is verry small, suggesting strong evidence against the null hypothesis.
    
(e) The R-squared value indicates that a large proportion of the variance in X is explained by Y.
    
Overall, the results suggest a strong positive relationship between Y and X. 

## (c) Relationship between (a) and (b)

1. The goodness of fit is idential in both models since the R-squared (or adjusted R-squared) and t-stat values are the same.

        ==> The strength of relationship is symmetric. Both regression capture the same underlying strong linear association since they are based on the same correlation.

2. The slope in (a) is 1.976 while the slope in (b) is 0.376. These slopes are not reciprocals. This is becasue each regression optimizes a differen criterion, so the line of best fit is different depending on which variable are treated as the dependent variable. In another word, regressing Y on X minimizes vertical errors in Y, while regression X on Y minizing verical error in X.

        ==> Slopes are asymmetric because one is scaled in units of Y per X and the other is in units of X per Y.


## (d) Formula Proof


In [ ]:
import numpy as np

# 1. Generate reproducible data
rng = np.random.default_rng(1)
x = rng.normal(size=100)
y = 2 * x + rng.normal(size=100)

n = len(x)

# 2. Compute the core sums
Sxy = np.sum(x * y)     # sum of cross-products
Sxx = np.sum(x**2)      # sum of squares of x
Syy = np.sum(y**2)      # sum of squares of y

# 3. Compute beta-hat and standard error (no intercept case)
beta_hat = Sxy / Sxx
RSE = Syy - (Sxy**2) / Sxx
SE_beta = np.sqrt(RSE / ((n - 1) * Sxx))

t_direct = beta_hat / SE_beta   # definition

# 4. Alternative identity form for t-stat
t_identity = np.sqrt(n - 1) * Sxy / np.sqrt(Sxx * Syy - Sxy**2)

# 5. Print both to confirm equality
print("Beta-hat:           ", beta_hat)
print("SE(beta-hat):       ", SE_beta)
print("t (direct formula): ", t_direct)
print("t (identity form):  ", t_identity)
print("Difference:         ", t_direct - t_identity)


Beta-hat:            1.9762423774420508
SE(beta-hat):        0.11694837274226332
t (direct formula):  16.898417063035094
t (identity form):   16.898417063035094
Difference:          0.0
